src/main/java/io/aiven/kafka/connect/common/output/SinkSchemaBuilder.java

```diff
@@ -0,0 +1,183 @@
+            }
+            return schemaFields.endRecord();
+        } else {
+            return tryUnwrapEnvelope(record);
+        }
+    }
+
+    private Schema tryUnwrapEnvelope(final SinkRecord record) {
+        // envelope can be disabled only in case of single field
+        final OutputField field = getFields().iterator().next();
+
+        final Schema schema = outputFieldSchema(field, record);
+        if (schema.getType() == Schema.Type.MAP) {
+            @SuppressWarnings("unchecked") final Map<String, Object> value =
+                (Map<String, Object>) record.value();
+            final SchemaBuilder.FieldAssembler<Schema> schemaFields =
+                SchemaBuilder
+                    .builder(getNamespace())
+                    .record("connector_records")
+                    .fields();
+            for (final Map.Entry<String, Object> entry : value.entrySet()) {
+                schemaFields.name(entry.getKey()).type(schema.getValueType()).noDefault();
+            }
+            return schemaFields.endRecord();
+        } else if (schema.getType() == Schema.Type.RECORD) {
+            return getAvroData().fromConnectSchema(record.valueSchema());
+        } else {
+            return SchemaBuilder
+                .builder(getNamespace())
+                .record("connector_records")
+                .fields()
+                .name(field.getFieldType().name)
+                .type(schema)
+                .noDefault()
+                .endRecord();
+        }
+    }
+
+    private Schema headersSchema(final SinkRecord record) {
+        if (record.headers().isEmpty()) {
+            return SchemaBuilder.builder().nullType();
+        }
+        org.apache.kafka.connect.data.Schema headerSchema = null;
+        for (final Header h : record.headers()) {
+            if (Objects.isNull(h.schema())) {
+                throw new DataException("Header " + h + " without schema");
+            }
+            if (Objects.isNull(headerSchema)) {
+                headerSchema = h.schema();
+            } else if (headerSchema.type() != h.schema().type()) {
+                throw new DataException("Header schema " + h.schema()
+                    + " is not the same as " + headerSchema);
+            }
+        }
+        return SchemaBuilder.map().values(avroData.fromConnectSchema(headerSchema));
+    }
+
+    protected Schema outputFieldSchema(final OutputField field, final SinkRecord record) {
+        switch (field.getFieldType()) {
+            case KEY:
+                return avroData.fromConnectSchema(record.keySchema());
+            case OFFSET:
+            case TIMESTAMP:
+                return SchemaBuilder.builder().longType();
+            case VALUE:
+                return avroData.fromConnectSchema(record.valueSchema());
+            case HEADERS:
+                return headersSchema(record);
+            default:
+                throw new ConnectException("Unknown field type " + field);
+        }
+    }
```

Comment:
Do we need to throw `DataException` and `ConnectException` here or could be the same type?